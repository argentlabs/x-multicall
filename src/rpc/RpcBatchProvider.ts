import DataLoader from "dataloader";
import { RpcProvider, type RpcProviderOptions, type RPC } from "starknet";
import type { DataLoaderOptions } from "../types";

type RpcRequest<T extends keyof RPC.Methods = keyof RPC.Methods> = {
  method: T;
  params: RPC.Methods[T]["params"];
};

export class RpcBatchProvider extends RpcProvider {
  private wait: number;
  private batchSize: number;
  private loader: DataLoader<RpcRequest, RPC.Response> | undefined;

  constructor({
    batchInterval,
    maxBatchSize,
    ...optionsOrProvider
  }: DataLoaderOptions & RpcProviderOptions) {
    super(optionsOrProvider);
    this.wait = batchInterval ?? 0;
    this.batchSize = maxBatchSize ?? 20;
    this.loader = new DataLoader(this.batchRequests.bind(this), {
      batch: true,
      maxBatchSize: this.batchSize,
      cache: false, // Set cache to false if you don't want caching, or omit this line to enable caching
      batchScheduleFn: (cb) => setTimeout(cb, this.wait),
    });
  }

  private async batchRequests(
    requests: readonly RpcRequest[]
  ): Promise<RPC.Response[]> {
    const body = requests.map(({ method, params }, i) => ({
      method,
      params: params ?? [],
      jsonrpc: "2.0",
      id: i,
    }));

    const response = await fetch(this.nodeUrl, {
      method: "POST",
      body: JSON.stringify(body),
      headers: this.headers as Record<string, string>,
    });

    if (!response.ok) {
      const data = await response.text();
      throw new Error(
        `Failed to fetch, status: ${response.status}, body:\n${JSON.stringify(
          data
        )}`
      );
    }

    const data: RPC.Response[] = await response.json().catch(async () => {
      const data = await response.text();
      throw new Error(
        `Failed to parse response as JSON, body:\n${JSON.stringify(data)}`
      );
    });

    const sortedData = data.sort((a, b) => a.id - b.id); // Sort the response to match the order of the requests

    return sortedData;
  }

  protected async fetchEndpoint<T extends keyof RPC.Methods>(
    method: T,
    params: RPC.Methods[T]["params"]
  ): Promise<RPC.Methods[T]["result"]> {
    const request = { method, params };
    const response = await (this.loader?.load(request) ??
      this.batchRequests([request]).then((res) => res[0]));
    const { error, result } = response;
    this.errorHandler(error);
    return result;
  }
}
