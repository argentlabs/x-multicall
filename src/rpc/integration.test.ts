import { Mock, afterEach, beforeAll, describe, expect, mock, test } from "bun:test";
import { RpcBatchProvider } from "./RpcBatchProvider";
import { filterError } from "../utils.test";

function getBatchProvider() {
  if (!process.env.TEST_RPC_PROVIDER) {
    throw new Error("TEST_RPC_PROVIDER env variable is not set");
  }

  return new RpcBatchProvider({
    nodeUrl: process.env.TEST_RPC_PROVIDER,
    batchInterval: 200,
    maxBatchSize: 20,
  });
}

describe("RpcBatchProvider", () => {
  let globalFetchMock: Mock<typeof fetch>;
  beforeAll(() => {
    const originalFetch = global.fetch;
    globalFetchMock = global.fetch = mock(originalFetch);
  });

  test("should return the correct result for one call", async () => {
    expect(globalFetchMock.mock.calls.length).toEqual(0);
    const provider = getBatchProvider();

    // wait 400ms to make sure the batch is sent
    await new Promise((resolve) => setTimeout(resolve, 400));
    expect(globalFetchMock.mock.calls.length).toEqual(1);

    const result = await provider.callContract({
      contractAddress: "0x49D36570D4E46F48E99674BD3FCC84644DDD6B96F7C741B1562B82F9E004DC7",
      entrypoint: "balanceOf",
      calldata: ["0xdeadbeef0"],
    });

    expect(result).toEqual(["0x0", "0x0"]);
    expect(globalFetchMock.mock.calls.length).toEqual(2);
  });

  test("should return the correct result for multiple calls with one request", async () => {
    expect(globalFetchMock.mock.calls.length).toEqual(0);
    const provider = getBatchProvider();

    // wait 400ms to make sure the batch is sent
    await new Promise((resolve) => setTimeout(resolve, 400));
    expect(globalFetchMock.mock.calls.length).toEqual(1);

    const responses = await Promise.all(
      new Array(4).fill(null).map((_, i) =>
        provider.callContract({
          contractAddress: "0x49D36570D4E46F48E99674BD3FCC84644DDD6B96F7C741B1562B82F9E004DC7",
          entrypoint: "balanceOf",
          calldata: [`0xdeadbeef${i}`],
        })
      )
    );

    expect(responses).toEqual([
      ["0x0", "0x0"],
      ["0x0", "0x0"],
      ["0x0", "0x0"],
      ["0x0", "0x0"],
    ]);
    expect(globalFetchMock.mock.calls.length).toEqual(2);
  });

  test("one call fails in a batch", async () => {
    const provider = getBatchProvider();

    // wait 400ms to make sure the batch is sent
    await new Promise((resolve) => setTimeout(resolve, 400));
    expect(globalFetchMock.mock.calls.length).toEqual(1);

    const responses = await Promise.allSettled(
      new Array(4).fill(null).map((_, i) => {
        if (i === 2) {
          return provider.callContract({
            contractAddress: `0x49D36570D4E46F48E99674BD3FCC84644DDD6B96F7C741B1562B82F9E004DC7`,
            entrypoint: "balance_of",
            calldata: [`0xdeadbeef${i}`],
          });
        }
        return provider.callContract({
          contractAddress: `0x49D36570D4E46F48E99674BD3FCC84644DDD6B96F7C741B1562B82F9E004DC7`,
          entrypoint: "balanceOf",
          calldata: [`0xdeadbeef${i}`],
        });
      })
    );

    expect(filterError(responses)).toMatchSnapshot();
    expect(globalFetchMock.mock.calls.length).toEqual(2);
  });

  afterEach(() => {
    globalFetchMock.mockClear();
  });
});
