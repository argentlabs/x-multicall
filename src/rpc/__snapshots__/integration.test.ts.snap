// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`RpcBatchProvider one call fails in a batch 1`] = `
[
  {
    "status": "fulfilled",
    "value": {
      "result": [
        "0x0",
        "0x0",
      ],
    },
  },
  {
    "status": "fulfilled",
    "value": {
      "result": [
        "0x0",
        "0x0",
      ],
    },
  },
  {
    "reason": [LibraryError: -32603: Internal error: Error in the called contract (0x049d36570d4e46f48e99674bd3fcc84644ddd6b96f7c741b1562b82f9e004dc7):
Error at pc=0:104:
Got an exception while executing a hint: Hint Error: Entry point EntryPointSelector(StarkFelt("0x035a73cd311a05d46deda634c5ee045db92f811b4e74bca4437fcb5302b7af33")) not found in contract.
Cairo traceback (most recent call last):
Unknown location (pc=0:1678)
Unknown location (pc=0:1664)
],
    "status": "rejected",
  },
  {
    "status": "fulfilled",
    "value": {
      "result": [
        "0x0",
        "0x0",
      ],
    },
  },
]
`;
